src/Chapter3.hs

```diff
@@ -1007,6 +1157,147 @@ contestants, and write a function that decides the outcome of a fight!
+
+castDefSpellKnight :: Knight -> (Knight, Attack) 
+castDefSpellKnight (Knight  h atk def) = (FightingKnight h h atk def 3,  noAttack) 
+castDefSpellKnight (FightingKnight h hMax atk def _) = (FightingKnight h hMax atk def 3, noAttack)
+
+
+data Monster = Monster
+    { monsterHealth    :: Health
+    , monsterAttack   :: Attack
+    }  | FledMonster
+    deriving (Show)
+
+fleeMonster :: Monster -> (Monster, Attack)
+fleeMonster _ = (FledMonster, MkAttack 0)
+
+attackMonster :: Monster -> (Monster,Attack)
+attackMonster m = (m, monsterAttack m)
+
+
+class Fighter a where
+    takeDamage :: a -> Attack -> a
+    defeated   :: a -> Bool
+    fled       :: a -> Bool
+
+instance Fighter Knight where
+    takeDamage :: Knight -> Attack -> Knight
+    takeDamage (Knight  h atk def) a =  takeDamage (FightingKnight h h atk def 0)  a
+    takeDamage (FightingKnight  h maxH atk def boostDef) attack = FightingKnight remainingHealth maxH atk def (max (boostDef-1) 0)
+              where reductionFactor = if boostDef > 0 then 2 else 1
+                    effectiveDmg = max (unAttack attack - (unDefense def) * reductionFactor) 0
+                    remainingHealth = MkHealth (max (unHealth h - effectiveDmg) 0)
+
+
+    defeated :: Knight -> Bool
+    defeated (Knight h _ _) = unHealth h <= 0
+    defeated (FightingKnight h _ _ _ _) = unHealth h <= 0
+
+    fled :: Knight -> Bool
+    fled _ = False
+
+
+instance Fighter Monster where
+    takeDamage :: Monster -> Attack -> Monster
+    takeDamage (Monster h a) (MkAttack atk) = Monster remainingHealth a
+                where remainingHealth = MkHealth (max (unHealth h - atk) 0)
+    
+    defeated :: Monster -> Bool
+    defeated m = unHealth (monsterHealth m) <= 0
+
+        
+    fled :: Monster -> Bool
+    fled FledMonster = True
+    fled _ = False
+
+rotateOne :: [a] -> [a]
+rotateOne l = take (length l) (drop 1 (cycle l))
+
+fight :: (Fighter a, Fighter b) =>  a -> [a -> (a, Attack)] -> b -> [b -> (b, Attack)] -> FightResult
+fight fighter1 actions1 fighter2 actions2 = go fighter1 actions1 fighter2 actions2 0
+  where 
+  go :: (Fighter a, Fighter b) =>  a -> [a -> (a, Attack)] -> b -> [b -> (b, Attack)] -> Int -> FightResult
+  go fighter1 actions1 fighter2 actions2 round 
+      | (length actions1) == 0  = FirstFighterDefeated
+      | (length actions2) == 0  = SecondFighterDefeated
+      | fled fighter1         = FirstFighterFled
+      | fled fighter2         = SecondFighterFled
+      | defeated fighter1     = FirstFighterDefeated
+      | defeated fighter2     = SecondFighterDefeated
+      | round > 100           = Draw --round limit (no infinte fights), fighters are too tired to continue ;)
+      | round `mod` 2 == 0    = go attakingFighter1 newActions1 defendingFighter2 actions2 (round + 1) 
+      | otherwise             = go defendingFighter1 actions1 attakingFighter2 newActions2 (round + 1) 
+          where action1 : _ = actions1
```

Comment:
Duplicated code where I had no clear idea how to write it only once, due to types `a` and `b` :/